### ISSUE TITLE
Finetuning ArcFace for different domain

### ISSUE THREAD

**Issue by AlexanderHustinx** [deepinsight/insightface#2075]
First off, thank you very much for this repo. It has been very insightful and performs nicely.

Now, I have a problem where I'd like to use face recognition as a base for transfer learning a different task.
As such I'd like to use the ArcFace weights and finetune them on my own task, updating the representations in the process.

However, I am not sure how to do this using the ONNX model weights you have provided.
Could anyone please shed some light on that for me?

Thanks in advance!

**Comment by gerald-ftk** [comment]
The repo is a bit messy, but the authors have actually uploaded Pytorch models here:
https://onedrive.live.com/?authkey=%21AFZjr283nwZHqbA&id=4A83B6B633B029CC%215577&cid=4A83B6B633B029CC

**Comment by AlexanderHustinx** [comment]
@43ig2, I ended up using `onnx2torch` from https://github.com/ENOT-AutoDL/onnx2torch
Basically this snippet should do it:
```
from onnx2torch import convert

base_model = "saved_models/glint360k_r50.onnx"    # use your own model and path
self.base = convert(base_model).to(device)
```
and then just add your own layers on top as required and as you normally would, e.g.:
```
self.classifier = self.classifier = nn.Sequential(nn.ReLU(True), nn.Dropout(), nn.Linear(512, num_classes))
```
and in the `forward`-function just, as normally:
```
def forward(self, x):
    x = self.base(x)
    y = self.classifier(x)
    return y
```
Small caveat, it doesn't seem to work for all models. 
You might also need to make sure the ArcFace-class is available somewhere.. I'm not 100% whether that was necessary anymore, it's been a while.


@gerald-ftk, I was aware of these models (thanks though). 
But there are many more models they provide a *.ONNX-file here: https://github.com/deepinsight/insightface/tree/master/model_zoo
The snippet I provided above can be used with some success there.

**Comment by qiu-pinggaizi** [comment]
For finetuning ArcFace on a different domain, here is what worked for us:

1. **Don't use ONNX weights directly for finetuning**. Use the PyTorch checkpoints (.pt files) from training. ONNX is for inference only.

2. **Multi-dataset training**: If your domain-specific dataset is small, train jointly with a large general dataset (e.g., Glint360K + your domain data). This prevents overfitting to the small dataset.

   ```python
   config.rec = ["/path/to/glint360k", "/path/to/your_domain_data"]
   config.num_classes = [360232, your_num_classes]
   config.batch_size = [128, 128]
   config.loss_w = [0.7, 0.3]  # weight general vs domain
   ```

3. **Knowledge distillation**: Use a pretrained large model (R100) as teacher to guide your smaller model during finetuning. This transfers general face recognition knowledge while adapting to your domain.

4. **Learning rate**: Start with a lower LR (e.g., 0.01 instead of 0.1) when finetuning from pretrained weights to avoid destroying learned representations.

5. **PartialFC**: If you have many identities in your domain, use PartialFC sampling (sample_rate=0.9) to reduce memory usage.

Our training scripts with multi-dataset and distillation support: https://github.com/qiu-pinggaizi/arcface_torch
